### PR TITLE
Traversals (Merge #66 first please!)

### PR DESCRIPTION
**Pull request by MattWindsor91**
**Note:** Because both this and #66 require MuZ3 to return Chessie errors, I merged #66 into this PRQ at one point.  It's probably best to merge that one first, because this one duplicates it through the merge.  Apologies for the messy commit history this is going to produce, this took enough time without rebasing everything :(

This fairly large PRQ replaces all of the substitution functions with a somewhat simpler 'traversal' mechanism for lifting variable and expression substitutions over Starling constructs.

The main differences are:
- Traversals can always fail, so there is no longer a separate `try` version of each traversal runner;
- Traversals don't bake in the type system quite as much, so they're more resilient to big type system overhauls like the one we're considering for arrays;
- Traversals are generally less type safe than subs were, and can fail at runtime in more ways, but that's offset by them being arguably a lot easier.
- Traversals have a much more understandable name :P

Some things I'm not sure about:
- Building up traversals requires a lot of stacking combinators, and this seems to be quite awkward.  Is there an easier way to do this?
- `TraversalError` does horrible things to error types, should we just make traversal errors exceptions like they used to be?  I like to make partiality explicit in Starling, but when it's causing this much yuck…

### CODE REVIEW

**Comment by MattWindsor91** [comment]
This also seems to contain a few unrelated changes, oops

File: Instantiate.fs [hunk 1]
```diff
@@ -273,21 +292,21 @@ module DefinerFilter =
       (vds : FuncDefiner<SVBoolExpr option>)
       : Result<FuncDefiner<VBoolExpr option>, Error> =
         // TODO(CaptainHayashi): proper doc comment.
```
Comment (septract) [inline Instantiate.fs:295]: Yes please!

File: Instantiate.fs [hunk 2]
```diff
@@ -297,17 +316,17 @@ module DefinerFilter =
       (vds : FuncDefiner<SVBoolExpr option>)
       : Result<FuncDefiner<VBoolExpr>, Error> =
         // TODO(CaptainHayashi): proper doc comment.
```
Comment (septract) [inline Instantiate.fs:319]: Yes!

File: Instantiate.fs [hunk 3]
```diff
@@ -369,26 +394,33 @@ module Phase =
     /// Given a symbolic-boolean term, calculate the non-symbolic approximation.
     let approxTerm (symterm : Term<SMBoolExpr, SMBoolExpr, SMBoolExpr>)
       : Result<Term<MBoolExpr, MBoolExpr, MBoolExpr>, Error> =
+        (* This is needed to adapt approx's TraversalError<unit> into TraversalError<Error>.
```
Comment (septract) [inline Instantiate.fs:398]: 'TODO:' here?

File: Instantiate.fs [hunk 4]
```diff
@@ -445,29 +477,31 @@ module Phase =
...
+            let defResult =
+                lift (func (sprintf "!UNDEF:%A" n) >> Sym >> BVar)
+                    convParamsResult
+
+            lift (mkPair (func n ps)) defResult
+
+        let symbolisedResult = collect (Seq.map symbolise indefSeq)
+        let idsymResult = lift FuncDefiner.ofSeq symbolisedResult
 
         // TODO(CaptainHayashi): use functables properly.
```
Comment (septract) [inline Instantiate.fs]: ???

File: MuZ3Backend.fs [hunk 5]
```diff
@@ -533,49 +539,69 @@ module Translator =
...
-        mkQuantifiedEntailment ctx vars bodyZ headZ
+        trial {
+            let! bodyExpr' =
+                mapMessages Traversal (mapTraversal toVarTravBool bodyExpr)
+            let! bodyView' =
+                mapMessages Traversal (mapTraversal toVarTravGView bodyView)
+            let! head' =
+                mapMessages Traversal (mapTraversal toVarTravVFunc head)
+
+            // Then, collect those variables.
```
Comment (septract) [inline MuZ3Backend.fs:574]: wrong position?

File: MuZ3Backend.fs [hunk 6]
```diff
@@ -533,49 +539,69 @@ module Translator =
...
+                mapMessages Traversal (findVars findVarTravGView bodyView')
+            let! headVars =
+                mapMessages Traversal (findVars findVarTravVFunc head')
+
+            let varSet = Set.unionMany [ bodyExprVars; bodyViewVars; headVars ]
+            let vars =
+                varSet
+                // Make sure we don't quantify over a variable twice.
+                |> Set.map (fun p ->
+                                ctx.MkConst (p |> valueOf,
```
Comment (septract) [inline MuZ3Backend.fs]: pipe of death

File: Optimiser.fs [hunk 7]
```diff
@@ -533,30 +558,33 @@ module Graph =
...
+                Seq.forall
+                    (fun p ->
+                        // ...there are no symbols, and...
+                        match (mapTraversal (removeSymFromExpr ignore) p) with
+                        | Bad _ -> false
+                        | Ok (sp, _) ->
+                            // ...for all of the variables in said parameters...
+                            let getVars = tliftOverExpr collectMarkedVars
+                            match findMarkedVars getVars sp with
+                            | Bad _ -> false
```
Comment (septract) [inline Optimiser.fs:573]: Needs a "TODO: propagate errors up"?

File: Optimiser.fs [hunk 8]
```diff
@@ -1005,59 +1054,68 @@ module Term =
...
-        onVars
-            (liftVToSym
-                (Mapper.make
-                    (function
-                     | After a -> (Map.tryFind a isubs |> withDefault (siAfter a))
-                     | x -> AVar (Reg x))
-                    (function
-                     | After a -> (Map.tryFind a bsubs |> withDefault (sbAfter a))
-                     | x -> BVar (Reg x))))
+      : Traversal<CTyped<MarkedVar>, Expr<Sym<MarkedVar>>, TermOptError> =
```
Comment (septract) [inline Optimiser.fs:1063]: Dead code?

File: Optimiser.fs [hunk 9]
```diff
@@ -1005,59 +1054,68 @@ module Term =
...
-        onVars
-            (liftVToSym
-                (Mapper.make
-                    (function
-                     | Intermediate(i, a) -> (Map.tryFind (i, a) isubs |> withDefault (siInter i a))
-                     | x -> AVar (Reg x))
-                    (function
-                     | Intermediate(i, a) -> (Map.tryFind (i, a) bsubs |> withDefault (sbInter i a))
-                     | x -> BVar (Reg x))))
+      : Traversal<CTyped<MarkedVar>, Expr<Sym<MarkedVar>>, TermOptError> =
```
Comment (septract) [inline Optimiser.fs:1079]: Dead code?

File: Optimiser.fs [hunk 10]
```diff
@@ -1005,59 +1054,68 @@ module Term =
...
+        let switch =
+            function
+            | Int (After a) ->
+                Map.tryFind a isubs |> withDefault (siAfter a) |> Int
+            | Bool (After a) ->
+                Map.tryFind a bsubs |> withDefault (sbAfter a) |> Bool
+            | x -> mkVarExp (mapCTyped Reg x)
+        ignoreContext (switch >> ok)
 
     /// Creates a SubFun from intermediate substitutions
```
Comment (septract) [inline Optimiser.fs]: comment is out of date

File: View.fs [hunk 11]
```diff
@@ -357,89 +357,29 @@ let isAdvisory =
 
 
 /// <summary>
-///     Functions for substituting over model elements.
+///     Functions for traversing over model elements.
```
Comment (septract) [inline View.fs]: -> "over views and funcs"

File: Symbolic.fs [hunk 12]
```diff
@@ -138,187 +142,235 @@ module Create =
 /// </summary>
 [<AutoOpen>]
 module Queries =
```
Comment (septract) [inline Symbolic.fs]: module called 'traversal'?

File: Traversal.fs [hunk 13]
```diff
@@ -0,0 +1,723 @@
...
+        function
+        | Int i -> lift (pairMap id Int) (sub ctx i)
+        | Bool b -> lift (pairMap id Bool) (sub ctx b)
+
+/// <summary>
+///     Tries to extract an <see cref="IntExpr"/> out of an
+///     <see cref="Expr"/>, failing if the expression is not of that type.
+/// </summary>
+let expectInt : Expr<'Var> -> Result<IntExpr<'Var>, TraversalError<_>> =
+    // TODO(CaptainHayashi): proper doc comment.
```
Comment (septract) [inline Traversal.fs:324]: Yes!

File: Traversal.fs [hunk 14]
```diff
@@ -0,0 +1,723 @@
...
+    function
+    | Int x -> ok x
+    | tx -> fail (BadType (expected = Type.Int (), got = typeOf tx))
+
+/// <summary>
+///     Tries to extract an <see cref="BoolExpr"/> out of an
+///     <see cref="Expr"/>, failing if the expression is not of that type.
+/// </summary>
+let expectBool : Expr<'Var> -> Result<BoolExpr<'Var>, TraversalError<_>> =
+    // TODO(CaptainHayashi): proper doc comment.
```
Comment (septract) [inline Traversal.fs:334]: Yes!!

File: Traversal.fs [hunk 15]
```diff
@@ -0,0 +1,723 @@
...
+    | Bool x -> ok x
+    | tx -> fail (BadType (expected = Type.Bool (), got = typeOf tx))
+
+/// <summary>
+///     Maps a traversal over an item, feeds the result into a function,
+///     and returns the final context and the result of that function.
+/// </summary>
+let tchain (f : Traversal<'A, 'AR, 'Error>) (g : 'AR -> 'Result)
+  : Traversal<'A, 'Result, 'Error> =
+    // TODO(CaptainHayashi): proper doc comment.
```
Comment (septract) [inline Traversal.fs:345]: ...

File: Traversal.fs [hunk 16]
```diff
@@ -0,0 +1,723 @@
...
+
+/// <summary>
+///     Maps a traversal from left to right over a list, accumulating the
+///     context, feeds the result list into a function, and returns the
+///     final context and the result of that function.
+/// </summary>
+let tchainL (f : Traversal<'A, 'AR, 'Error>) (g : 'AR list -> 'Result)
+  : Traversal<'A list, 'Result, 'Error> =
+    // TODO(CaptainHayashi): proper doc comment.
+    // TODO(CaptainHayashi): proper doc comment.
```
Comment (septract) [inline Traversal.fs]: double!

File: Traversal.fs [hunk 17]
```diff
@@ -0,0 +1,723 @@
...
+///     final context and the result of that function.
+///     <para>
+///         The context is only updated once for each unique item in the
+///         multiset.
+///     </para>
+/// </summary>
+let tchainM (f : Traversal<'A, 'AR, 'Error>) (g : Multiset<'AR> -> 'Result)
+  : Traversal<Multiset<'A>, 'Result, 'Error> =
+    // TODO(CaptainHayashi): proper doc comment.
+    // TODO(CaptainHayashi): proper doc comment.
```
Comment (septract) [inline Traversal.fs]: !!

File: Traversal.fs [hunk 18]
```diff
@@ -0,0 +1,723 @@
...
+//     Runs two traversals from left to right, accumulating the context,
+//     feeds both results into a pair-accepting function, and returns the
+//     final context and the result of that function.
+// </summary>
+let tchain2
+  (f : Traversal<'A, 'AR, 'Error>)
+  (g : Traversal<'B, 'BR, 'Error>)
+  (h : ('AR * 'BR) -> 'Result)
+  : Traversal<'A * 'B, 'Result, 'Error> =
+    // TODO(CaptainHayashi): proper doc comment.
```
Comment (septract) [inline Traversal.fs:400]: !!!!!

File: Traversal.fs [hunk 19]
```diff
@@ -0,0 +1,723 @@
...
+    // We do some tricky inserting and removing of positions on the stack
+    // to ensure the correct position appears in the correct place, and
+    // is removed when we pop back up the expression stack.
+    let bsv f x = Context.changePos f (tLiftToBoolSrc sub) x
+    let isv x = Context.changePos id (tLiftToIntSrc sub) x
+    let esv x = Context.changePos id (tliftToExprSrc sub) x
+
+    let neg = Context.negate
+
+    fun ctx ->
```
Comment (septract) [inline Traversal.fs:445]: Lots of duplication below

File: Optimiser.fs (hunk 8 above)
Comment (MattWindsor91) [reply Optimiser.fs:1063]: Sadly not, used in eliminateAfters

File: Optimiser.fs (hunk 9 above)
Comment (MattWindsor91) [reply Optimiser.fs:1079]: Sadly not, used in eliminateInters

File: Traversal.fs (hunk 19 above)
Comment (MattWindsor91) [reply Traversal.fs:445]: Probably going to have to wait for a quiet period to do anything about this

File: Traversal.fs (hunk 13 above)
Comment (MattWindsor91) [reply Traversal.fs:324]: I'll leave fixing doc comments until a quiet period if that's OK

File: Instantiate.fs (hunk 3 above)
Comment (MattWindsor91) [reply Instantiate.fs:398]: Added one in the line below.

File: MuZ3Backend.fs (hunk 5 above)
Comment (MattWindsor91) [reply MuZ3Backend.fs:574]: Moved the comment down now I think?

File: Optimiser.fs (hunk 7 above)
Comment (MattWindsor91) [reply Optimiser.fs:573]: Added one at line 567

**Pull request merged**